File: python_cpp/uPIMulator_backend/src/simulator/sram/iram.cc
#include "simulator/sram/iram.h"

#include "encoder/instruction_encoder.h"

namespace upmem_sim::simulator::sram {

IRAM::IRAM() {
  address_ = new abi::word::InstructionAddressWord();
  address_->set_value(util::ConfigLoader::iram_offset());

  size_ = util::ConfigLoader::iram_size();

  assert(address() % abi::word::InstructionWord().size() == 0);
  assert(size_ % abi::word::InstructionWord().size() == 0);

  cells_.resize(num_instruction_words());
  for (int i = 0; i < num_instruction_words(); i++) {
    cells_[i] = new abi::word::InstructionWord();
  }
}

IRAM::~IRAM() {
  delete address_;

  for (int i = 0; i < size_ / num_instruction_words(); i++) {
    delete cells_[i];
  }
}

abi::instruction::Instruction *IRAM::read(Address address) {
  encoder::ByteStream *byte_stream = cells_[index(address)]->to_byte_stream();
  abi::instruction::Instruction *instruction =
      encoder::InstructionEncoder::decode(byte_stream);
  delete byte_stream;
  return instruction;
}

void IRAM::write(Address address, encoder::ByteStream *byte_stream) {
  cells_[index(address)]->from_byte_stream(byte_stream);
}

int IRAM::index(Address address) {
  assert(address >= this->address());
  assert(address + abi::word::InstructionWord().size() <=
         this->address() + size_);
  assert((address - this->address()) % abi::word::InstructionWord().size() ==
         0);

  return static_cast<int>((address - this->address()) /
                          abi::word::InstructionWord().size());
}

}  // namespace upmem_sim::simulator::sram

File: python_cpp/uPIMulator_frontend/src/iss/sram/iram.py
from typing import List

from abi.word.instruction_address_word import InstructionAddressWord
from abi.word.instruction_word import InstructionWord
from abi.word.representation import Representation
from util.config_loader import ConfigLoader


class IRAM:
    def __init__(self):
        self._address: InstructionAddressWord = InstructionAddressWord()
        self._address.set_value(ConfigLoader.iram_offset())

        self._instruction_words: List[InstructionWord] = [
            InstructionWord() for _ in range(ConfigLoader.iram_size() // InstructionWord().size())
        ]

    def address(self) -> int:
        return self._address.value(Representation.UNSIGNED)

    def read(self, address: int) -> InstructionWord:
        index = self._index(address)
        instruction_word = InstructionWord()
        instruction_word.set_value(self._instruction_words[index].value(Representation.UNSIGNED))
        return instruction_word

    def write(self, address: int, instruction_word: InstructionWord) -> None:
        index = self._index(address)
        self._instruction_words[index].set_value(instruction_word.value(Representation.UNSIGNED))

    def _index(self, address: int) -> int:
        assert address % InstructionWord().size() == 0
        index = (address - self.address()) // InstructionWord().size()
        assert 0 <= index < len(self._instruction_words)

        return index
